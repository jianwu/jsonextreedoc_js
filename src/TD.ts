import { TDJSONWriterOption, TDJSONWriter, TDObjectCoder, TDJSONParser } from '.';
import { TDObjectCoderOption } from '.';

export class TDEncodeOption {
  constructor(
    public coderOption: TDObjectCoderOption = new TDObjectCoderOption(),
    public jsonOption: TDJSONWriterOption = new TDJSONWriterOption(),
  ) {}

  public setCodeOption(codeOpt : TDObjectCoderOption) {
    this.coderOption = codeOpt;
    return this;
  }

  public setJsonOption(jsonOpt : TDJSONWriterOption) {
    this.jsonOption = jsonOpt;
    return this;
  }
}

export class TDDecodeOption {}
/**
 * A group of convenient methods similar to JSON
 */
export default class TD {
  public static stringify(obj: any, opts: string | TDEncodeOption = '  '): string {
    const opt = typeof opts === 'string' ? new TDEncodeOption() : opts;
    if (typeof opts === 'string') 
      opt.jsonOption.setIndentStr(opts);

    return TDJSONWriter.get().writeAsString(
      TDObjectCoder.get().encode(obj, opt.coderOption),
      opt.jsonOption,
    );
  }

  public static parse(str: string): any {
    return TDJSONParser.parse(str).toObject(false);
  }
}
