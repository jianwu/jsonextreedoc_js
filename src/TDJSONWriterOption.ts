export default class TDJSONWriterOption {
  public indentFactor = 0;
  public alwaysQuoteName = true;
  public quoteChar = '"';
  /** @internal */
  public indentStr = "";  // Used internally

  public setIndentFactor(indentFactor: number): TDJSONWriterOption {
    this.indentFactor = indentFactor;
    this.indentStr = ' '.repeat(this.indentFactor);
    return this;
  }

  public setAlwaysQuoteName(alwaysQuoteName : boolean) {
    this.alwaysQuoteName = alwaysQuoteName;
    return this;
  }

  public setQuoteChar(quoteChar: string) {
    this.quoteChar = quoteChar;
    return this;
  }
}