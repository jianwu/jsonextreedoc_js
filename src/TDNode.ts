import { Bookmark } from './Bookmark';
import { TreeDoc } from './TreeDoc';
import { TDPath, Part, PathPartType } from './TDPath';
import { StringUtil } from './core/StringUtil';
import { TDNodeProxyHandler } from './TDNodeProxyHandler';


export enum TDNodeType {
  MAP,
  ARRAY,
  SIMPLE,
}

export type ValueType = string | number | boolean | null | undefined;

const KEY_REF = '$ref';

export class TransientData {
  hash?: number;
  str?: string;
  obj?: any;
  proxy?: any;
  [key: string]: any;
}

export class TDNode {
  public static readonly ID_KEY = "$id";
  public static readonly REF_KEY = "$ref";
  
  public parent?: TDNode;
  public type = TDNodeType.SIMPLE;
  /** The value of the node, only available for leave node */
  private mValue: ValueType = null;
  /** Children of node. Use List instead of Map to avoid performance overhead of HashMap for small number of elements */
  public children?: TDNode[];
  /** Start position in the source */
  public start?: Bookmark;
  /** Length of this node in the source */
  public end?: Bookmark;
  /** indicate this node is a deduped Array node for textproto which allows duplicated keys */
  public deduped = false;

  // transient properties
  public readonly tData = new TransientData();

  // Create a root node if parent is undefined
  public constructor(public doc: TreeDoc, public key?: string) {}

  public clone(): TDNode {
    const result = new TDNode(this.doc, this.key).setType(this.type).setValue(this.value);
    result.parent = this.parent;
    result.children = this.children;
    result.start = this.start;
    result.end = this.end;
    result.deduped = this.deduped;
    return result;
  }

  public cloneOfSimpleType(value: ValueType): TDNode { return new TDNode(this.doc, this.key).setParent(this.parent).setType(TDNodeType.SIMPLE).setValue(value); }

  public setParent(parent?: TDNode): TDNode { this.parent = parent; return this.touch(); }
  public setKey(key?: string): TDNode { this.key = key; return this.touch(); }
  public setValue(val?: ValueType): TDNode { this.mValue = val; return this.touch(); }
  public get value() { return this.mValue; }
  public set value(value: ValueType) { this.mValue = value; }
  public setType(type: TDNodeType): TDNode { this.type = type; return this; }
  public setStart(start: Bookmark) { this.start = start; return this;}
  public setEnd(end: Bookmark) { this.end = end; return this;}

  public createChild(name?: string): TDNode {
    const childIndex = this.indexOf(name);
    if (childIndex < 0) {
      const cnode = new TDNode(this.doc, name);
      this.addChild(cnode);
      return cnode;
    }

    // special handling for textproto due to it's bad design that allows duplicated keys
    let existNode = this.children![childIndex].clone();

    if (!existNode.deduped) {
      const listNode = new TDNode(this.doc, name);
      listNode.parent = this;
      listNode.deduped = true;
      listNode.type = TDNodeType.ARRAY;

      this.children![childIndex] = listNode;
      existNode.key = '0';
      listNode.addChild(existNode);
      listNode.start = existNode.start; // Reuse first node's start and length
      listNode.end = existNode.end;
      existNode = listNode;
    }

    return existNode.createChild();
  }

  public addChild(node: TDNode) {
    if (!this.children) 
      this.children = [];
    node.parent = this;
    node.doc = this.doc;
    if (node.key == null)  // Assume it's array element
      node.key = "" + this.getChildrenSize();
    this.children.push(node);
    return this.touch();
  }

  public getChild(name: string | number): TDNode | null {
    if (typeof name === 'string')
      name = this.indexOf(name);
    return this.hasChildren() && name >= 0 ? this.children![name] : null;
  }

  public indexOf(name?: string): number {
    // VUETIPS: When VueJS instrument this object, it will generate getter to register the dep-graph.
    // very call to getter could be very heavy if the number of children is huge e.g. > 10000.
    // If we put the this.children inside for loop, it will cause O(n^2) problem.
    // so we have to cache it outside the for loop.
    // Relevant code: reactiveGetter (vue.runtime.esm.js?2b0e:1031)
    // TODO: add index when the node number is huge
    const children = this.children;
    if (!children || name == null)
      return -1;
    for (let i = 0; i < children.length; i++)
      if (name === children[i].key)
        return i;
    return -1;
  }

  public getChildValue(name: string): ValueType {
    const cn = this.getChild(name);
    return cn == null ? null : cn.value;
  }

  public hasChildren() {
    return this.children && this.children.length > 0;
  }
  public getChildrenSize() {
    return !this.children ? 0 : this.children.length;
  }

  public getValueByPath(path: TDPath | string): ValueType {
    const cn = this.getByPath(path);
    return cn && cn.value;
  }

  /** If noNull is true, it will return the last matched node */
  public getByPath(path: TDPath | string | string[], noNull = false, idx = 0): TDNode | null {
    if (!(path instanceof TDPath))
      path = TDPath.parse(path);

    if (idx === path.parts.length)
      return this;

    const next = this.getNextNode(path.parts[idx]);
    if (next == null)
      return noNull ? this : null;

    return next.getByPath(path, noNull, idx + 1);
  }

  public getNextNode(part: Part): TDNode | null {
    switch (part.type) {
      case PathPartType.ROOT:
        return this.doc.root;
      case PathPartType.CHILD_OR_ID:
        return this.getChild(part.key!) || this.doc.idMap[part.id!];
      case PathPartType.RELATIVE:
        return this.getAncestor(part.level!);
      case PathPartType.CHILD:
        return this.getChild(part.key!);
      default:
        return null; // Impossible
    }
  }

  public getAncestor(level: number): TDNode | null {
    let result: TDNode | null = this;
    for (let i = 0; i < level && result != null; i++, result = result.parent || null)
      ;
    return result;
  }

  public foreach(action: (n: TDNode) => void) {
    action(this);
    if (this.children != null)
      this.children.forEach(n => n.foreach(action));
    return this;
  }

  public isRoot() {
    return !this.parent;
  }

  /** JS specific logic */
  public toObject(includePosition = false, useCache = true): any {
    if (this.tData.obj !== undefined && useCache)
      return this.tData.obj;

    const $ = {
      start: this.start,
      end: this.end,
    };

    switch (this.type) {
      case TDNodeType.SIMPLE:
        return this.value;
      case TDNodeType.MAP: {
        this.tData.obj = includePosition ? { $ } : {};

        const refVal = this.getChildValue(KEY_REF);
        if (typeof refVal === 'string') {
          const target = this.getByPath(refVal);
          if (target == null)
            throw new Error(`Reference is not found: ref:${refVal}; current Node:${this.pathAsString}`);
          this.tData.obj = target.toObject(includePosition);
        } else {
          if (this.children) 
            this.children.forEach(c => c.key && (this.tData.obj[c.key] = c.toObject(includePosition)));
        }
        return this.tData.obj;
      }
      case TDNodeType.ARRAY: {
        this.tData.obj = [];
        if (includePosition)
          (this.tData.obj as any).$ = $;
        if (this.children)
          this.children.forEach(c => this.tData.obj.push(c.toObject(includePosition)));
        return this.tData.obj;
      }
      default:
        throw new Error('Unknown type');
    }
  }

  public toProxy(useCache = true): TDNode {
    if (useCache && this.tData.proxy)
      return this.tData.proxy;
    return this.tData.proxy = new Proxy(this, new TDNodeProxyHandler(useCache));
  }

  public get pathAsString() {
    return '/' + this.path.join('/');
  }
  public get path(): string[] {
    return this.parent ? [...this.parent.path, this.key!] : [];
  }
  public isLeaf() {
    return this.getChildrenSize() === 0;
  }

  private touch(): TDNode {
    this.tData.hash = undefined;
    this.tData.str = undefined;
    this.tData.proxy = undefined;
    this.tData.obj = undefined;
    if (this.parent != null)
      this.parent.touch();
    return this;
  }

  public toString() {
    if (this.tData.str === undefined)
      this.tData.str = this.toStringInternal('');
    return this.tData.str;
  }

  /** method specific for JSON.stringify() */
  public toJSON() { return toString(); }

  public toStringInternal(sb: string, includeRootKey = true, includeReservedKeys = true, limit = 100000) {
    if (this.parent != null && this.parent.type === TDNodeType.MAP && includeRootKey)
      sb += this.key + ': ';


    if (this.value !== null && this.value !== undefined) {
      if (typeof this.value !== 'string') {
        sb += this.value;
      } else {
        let str = StringUtil.cEscape(this.value, '\'')!;
        const remainLen = limit - sb.length;
        if (str.length > remainLen)
          str = str.substring(0, remainLen) + "...";
        sb += '\'' + str + '\'';
      }
    }

    if (this.children == null)
      return sb;

    sb += this.type === TDNodeType.ARRAY ? '[' : '{';
    for (const n of this.children) {
      if (!includeReservedKeys && n.key && n.key.startsWith("$"))
        continue;

      if (sb.length > limit) {
        sb += '...';
        break;
      }
  
      sb = n.toStringInternal(sb, true, includeReservedKeys, limit);      
      if (n !== this.children.slice(-1)[0])
        sb += ", ";

    }
    sb += this.type === TDNodeType.ARRAY ? ']' : '}';
    return sb;
  }

  public freeze() {
    const children = this.children;
    if (children) {
      for (const c of children) 
        c.freeze();
    }
    Object.freeze(this.start);
    Object.freeze(this.end);
    Object.freeze(this.doc);
    Object.freeze(this);
  }
}
