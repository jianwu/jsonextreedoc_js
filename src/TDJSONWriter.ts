import TDJSONWriterOption from "./TDJSONWriterOption";
import TDNode, { TDNodeType } from "./TDNode";
import StringBuilder from "./StringBuilder";
import Appendable from "./Appendable";
import StringUtil from "./StringUtil";

export default class TDJSONWriter {
  public static readonly instance = new TDJSONWriter();
  public static get() { return TDJSONWriter.instance; }

  public writeAsString(node: TDNode, opt = new TDJSONWriterOption()): string {
    const out = new StringBuilder();
    this.write(out, node, opt);
    return out.toString();
  }

  public write(out: Appendable, node: TDNode, opt: TDJSONWriterOption, indentStr = ""): void {
    if (!node) {
      out.append("null");
      return;
    }

    const isCompact = opt.indentFactor === 0;
    let childIndentStr = "";
    if (!isCompact)
      childIndentStr = indentStr + opt.indentStr;

    switch (+node.type) {
      case TDNodeType.MAP:
        this.writeMap(out, node, opt, indentStr, childIndentStr);
        return;
      case TDNodeType.ARRAY:
        this.writeArray(out, node, opt, indentStr, childIndentStr);
        return;
      default:
        this.writeSimple(out, node, opt);
    }
  }

  private writeMap(out: Appendable, node: TDNode, opt: TDJSONWriterOption, indentStr: string, childIndentStr: string): void {
    out.append('{');
    if (node.children != null) {
      for (let i = 0; i < node.children.length; i++){
        const cn = node.children[i];
        if (opt.indentFactor > 0) {
          out.append('\n');
          out.append(childIndentStr);
        }
        if (!StringUtil.isJavaIdentifier(cn.key) || opt.alwaysQuoteName)  // Quote the key in case  it's not valid java identifier
          this.writeQuotedString(out, cn.key as string, opt.quoteChar);
        else
          out.append(cn.key as string);  // Map key will never be null
        out.append(":");
        this.write(out, cn, opt, childIndentStr);
        if (i < node.children.length - 1) // No need "," for last entry
          out.append(",");
      }

      if (opt.indentFactor > 0 && node.children.length > 0) {
        out.append('\n');
        out.append(indentStr);
      }
    }

    out.append('}');
  }

  private writeArray(out: Appendable, node: TDNode, opt: TDJSONWriterOption, indentStr: string, childIndentStr: string): void {
    out.append('[');
    if (node.children != null) {
      for (let i = 0; i < node.children.length; i++) {
        const cn = node.children[i];
        if (opt.indentFactor > 0) {
          out.append('\n');
          out.append(childIndentStr);
        }
        this.write(out, cn, opt, childIndentStr);
        if (i < node.children.length - 1) // No need "," for last entry
          out.append(",");
      }

      if (opt.indentFactor > 0 && node.children.length > 0) {
        out.append('\n');
        out.append(indentStr);
      }
    }

    out.append(']');
  }

  private writeSimple(out: Appendable, node: TDNode, opt: TDJSONWriterOption): void {
    if (typeof(node.value) === 'string') {
      this.writeQuotedString(out, node.value as string, opt.quoteChar);
      return;
    }

    out.append(node.value + "");
  }

  private writeQuotedString(out: Appendable, str: string, quoteChar: string): void {
    out.append(quoteChar);
    out.append(StringUtil.cEscape(str, quoteChar) as string);
    out.append(quoteChar);
  }
}
