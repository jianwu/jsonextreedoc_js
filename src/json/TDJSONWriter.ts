import TDJSONWriterOption from './TDJSONWriterOption';
import TDNode, { TDNodeType } from '../TDNode';
import StringBuilder from '../core/StringBuilder';
import Appendable from '../core/Appendable';
import StringUtil from '../core/StringUtil';
import { TreeDoc } from '..';
import LangUtil from '../core/LangUtil';

export default class TDJSONWriter {
  public static readonly instance = new TDJSONWriter();
  public static get() {
    return TDJSONWriter.instance;
  }

  public static writeAsString(node: TDNode, opt: Partial<TDJSONWriterOption> = {}): string {
    return TDJSONWriter.get().writeAsString(node, opt);
  }

  public static write(out: Appendable, node: TDNode, opt: Partial<TDJSONWriterOption> = {}, indentStr = ''): Appendable {
    return TDJSONWriter.get().write(out, node, opt, (indentStr = ''));
  }

  public writeAsString(node: TDNode, opt: Partial<TDJSONWriterOption> = {}): string {
    const out = new StringBuilder();
    this.write(out, node, opt);
    return out.toString();
  }

  public write(out: Appendable, node: TDNode, option: Partial<TDJSONWriterOption> = {}, indentStr = ''): Appendable {
    if (!node)
      return out.append('null');

    const opt = LangUtil.mergeDeep(new TDJSONWriterOption(), option);
    let childIndentStr = '';
    if (opt.hasIndent())
      childIndentStr = indentStr + opt.indentStr;

    switch (node.type) {
      case TDNodeType.MAP:
        return this.writeMap(out, node, opt, indentStr, childIndentStr);
      case TDNodeType.ARRAY:
        return this.writeArray(out, node, opt, indentStr, childIndentStr);
      default:
        return this.writeSimple(out, node, opt);
    }
  }

  private writeMap(out: Appendable, node: Readonly<TDNode>, opt: TDJSONWriterOption, indentStr: string, childIndentStr: string): Appendable {
    out.append('{');
    if (node.children != null) {
      for (let i = 0; i < node.getChildrenSize(); i++) {
        const cn = opt.applyFilters(node.children[i]);
        if (cn == null)
          continue;

        if (opt.hasIndent()) {
          out.append('\n');
          out.append(childIndentStr);
        }
        if (!StringUtil.isJavaIdentifier(cn.key) || opt.alwaysQuoteName)
          // Quote the key in case  it's not valid java identifier
          this.writeQuotedString(out, cn.key as string, opt.quoteChar);
        else out.append(cn.key as string); // Map key will never be null
        out.append(':');
        this.write(out, cn, opt, childIndentStr);
        if (i < node.getChildrenSize() - 1)
          // No need "," for last entry
          out.append(',');
      }

      if (opt.hasIndent() && node.hasChildren()) {
        out.append('\n');
        out.append(indentStr);
      }
    }

   return  out.append('}');
  }

  private writeArray(out: Appendable, node: Readonly<TDNode>, opt: TDJSONWriterOption, indentStr: string, childIndentStr: string): Appendable {
    out.append('[');
    if (node.children != null) {
      for (let i = 0; i < node.getChildrenSize(); i++) {
        const cn = node.children[i];
        if (opt.hasIndent()) {
          out.append('\n');
          out.append(childIndentStr);
        }
        this.write(out, cn, opt, childIndentStr);
        if (i < node.getChildrenSize() - 1)
          // No need "," for last entry
          out.append(',');
      }

      if (opt.hasIndent() && node.children.length > 0) {
        out.append('\n');
        out.append(indentStr);
      }
    }

    return out.append(']');
  }

  private writeSimple(out: Appendable, node: TDNode, opt: TDJSONWriterOption): Appendable {
    const value = node.value;
    if (typeof value === 'string')
      return this.writeQuotedString(out, value as string, opt.quoteChar);

    return out.append(value + '');
  }

  private writeQuotedString(out: Appendable, str: string, quoteChar: string): Appendable {
    out.append(quoteChar);
    out.append(StringUtil.cEscape(str, quoteChar) as string);
    return out.append(quoteChar);
  }
}
