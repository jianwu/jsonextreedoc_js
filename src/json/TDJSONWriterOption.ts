import { TDNode } from "..";
import { ValueType } from "../TDNode";
import NodeFilter from "./NodeFilter";

export default class TDJSONWriterOption {
  private mIndentFactor = 0;
  public alwaysQuoteName = true;
  public quoteChar = '"';
  /** @internal */
  public indentStr = ''; // Used internally
  
  /** Node mapper, if it returns null, node will be skipped */
  public nodeFilters: NodeFilter[] = [];

  public set indentFactor(indentFactor: number) {
    this.mIndentFactor = indentFactor;
    this.indentStr = ' '.repeat(this.mIndentFactor);
  }

  public get indentFactor() { return this.mIndentFactor}

  public setIndentFactor(indentFactor: number) {
    this.indentFactor = indentFactor;
    return this;
  }

  public setIndentStr(str: string) {
    this.indentStr = str;
    return this;
  }

  public setAlwaysQuoteName(alwaysQuoteName: boolean) {
    this.alwaysQuoteName = alwaysQuoteName;
    return this;
  }

  public setQuoteChar(quoteChar: string) {
    this.quoteChar = quoteChar;
    return this;
  }

  public hasIndent() {
    return this.indentStr.length > 0;
  }

  public applyFilters(n: TDNode): TDNode | undefined {
    let result: TDNode | undefined = n;
    for (const f of this.nodeFilters){
      if (result == null)
        break;
      result = f.apply(result);
    }
    return result;
  }

  public addNodeFilter(...filters: NodeFilter[]) {
    for (const f of filters)
      this.nodeFilters.push(f);
    return this;
  }
}
