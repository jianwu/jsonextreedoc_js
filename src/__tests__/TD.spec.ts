import TDJSONWriter from '../json/TDJSONWriter';
import TDJSONWriterOption from '../json/TDJSONWriterOption';

import TD, { TDEncodeOption } from '../TD';
import { TDNodeType } from '../TDNode';

class TestObject {
  constructor(public title: string) {}
  nullObj: null;
  functionObj(num: number[]) {
    console.log(num);
  }
}

const commonObj = new TestObject('common');

const obj: any = {
  num: 10,
  obj: {
    str: '123',
    bool: true,
    date: new Date('2019-12-16T17:34:45.024Z'),
    nestObj: commonObj,
    cyclic: null,
    specialArray: [10, undefined, (a: number) => {;}, Symbol('')],
  },
  obj1: commonObj,
};

const commonObjStr = `{
  $type:'TestObject',
  title:'common'
}`;

// ES6 doesn't work
const commonObjConstructorEs5 = `{
  $type:'TestObject',
  functionObj:'function (num) {\\n        console.log(num);\\n    }'
}`

const commonObjConstructorEs6 = `{
  $type:'TestObject'
}`

const objStr = `{
  "num":10,
  "obj":{
    "str":"123",
    "bool":true,
    "date":"2019-12-16T17:34:45.024Z",
    "nestObj":{
      "title":"common",
      "$id":1
    },
    "cyclic":{
      "$ref":"../../"
    },
    "specialArray":[
      10,
      null,
      null,
      null
    ]
  },
  "obj1":{
    "$ref":"#1"
  },
  "arrayRef":{
    "$ref":"/obj/specialArray"
  }
}`;
obj.obj.cyclic = obj;
obj.arrayRef = obj.obj.specialArray;

describe('TD', () => {
  test('stringify with Options', () => {
    expect(TD.stringify(commonObj, {
      coderOption: {
        showType: true
      },
      jsonOption: { 
        alwaysQuoteName: false, 
        quoteChar: "'", 
        indentFactor: 2}
    })).toBe(commonObjStr);
  });

  test('stringify with showFunction', () => {
    const opt = new TDEncodeOption();
    opt.jsonOption
        .setAlwaysQuoteName(false)
        .setQuoteChar("'")
        .setIndentFactor(2);
    opt.coderOption
        .setShowType(true)
        .setShowFunction(true);
    // Not sure why Object.keys doesn't return any keys for this particular test case for ES6
    expect(TD.stringify(commonObj.constructor.prototype, opt)).toBe(commonObjConstructorEs5);
  });

  test('stringify cyclic without options', () => {
    console.log(TD.stringify(obj));
    expect(TD.stringify(obj, "  ")).toBe(objStr);
  });

  test('parse', () => {
    const o = TD.parse(objStr);
    const str = TD.stringify(o, "  ");
    expect(str).toEqual(objStr);
    console.log(str);
    expect(o.obj1).toBe(o.obj.nestObj);
    expect(o.obj.cyclic).toBe(o);
    expect(o.arrayRef).toBe(o.obj.specialArray);
  });

  test('parseWithOption', () => {
    const o = TD.parse("a,b,c", { defaultRootType: TDNodeType.ARRAY });
    expect(o).toEqual(["a", "b", "c"]);
  });
});
