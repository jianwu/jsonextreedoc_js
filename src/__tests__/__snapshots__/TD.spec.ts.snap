// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TD stringifyTDNode 1`] = `"{\\"doc\\":{\\"idMap\\":{},\\"root\\":{\\"$ref\\":\\"../../\\"},\\"$id\\":1},\\"key\\":\\"root\\",\\"type\\":0,\\"deduped\\":false,\\"tData\\":{},\\"start\\":{\\"line\\":0,\\"col\\":0,\\"pos\\":0},\\"children\\":[{\\"doc\\":{\\"$ref\\":\\"#1\\"},\\"key\\":\\"a\\",\\"type\\":2,\\"mValue\\":1,\\"deduped\\":false,\\"tData\\":{},\\"parent\\":{\\"$ref\\":\\"../../../\\"},\\"start\\":{\\"line\\":0,\\"col\\":3,\\"pos\\":3},\\"end\\":{\\"line\\":0,\\"col\\":4,\\"pos\\":4}},{\\"doc\\":{\\"$ref\\":\\"#1\\"},\\"key\\":\\"b\\",\\"type\\":2,\\"mValue\\":2,\\"deduped\\":false,\\"tData\\":{},\\"parent\\":{\\"$ref\\":\\"../../../\\"},\\"start\\":{\\"line\\":0,\\"col\\":8,\\"pos\\":8},\\"end\\":{\\"line\\":0,\\"col\\":9,\\"pos\\":9}}],\\"end\\":{\\"line\\":0,\\"col\\":10,\\"pos\\":10}}"`;
